plone/app/z3cform/templates/text_input.pt

```diff
@@ -2,51 +2,5 @@
       xmlns:tal="http://xml.zope.org/namespaces/tal"
       tal:omit-tag=""
 >
-  <input class=""
-         id=""
-         accesskey=""
-         alt=""
-         lang=""
-         maxlength=""
-         name=""
-         size=""
-         style=""
-         tabindex=""
-         title=""
-         type="text"
-         value=""
-         tal:attributes="
-           id view/id;
-           name view/name;
-           required python:view.required and 'required' or None;
-           class python:'form-control {0}{1}'.format(view.klass, view.error and ' is-invalid' or '');
```

Comment:
This class logic has to be implemented in `attributes` in the widgets view.

Reply:
The `class` attribute is handled specially with the `addClass()` method ... this is to enable stacked css classes for subclassed widgets with the attribute `klass`. At least that's how I understood it.

Reply:
right, super. then it's fine!